(sentry-metrics): Metrics indexer consumer

MeredithAnya opened the pull request:
**Metrics Indexer Consumer:**
Messages produced by Relay into the `ingest-metrics` topic have a metric `name` along with any number of tag key, value string pairs associated with the metric. The snuba topic `snuba-metrics` (which be changed [here](https://github.com/getsentry/snuba/blob/e6bfca988eb2d85c5d41eddbc004b9000128c9c8/snuba/utils/streams/topics.py#L12)) expects the integers instead of strings. 

The indexer (which will be implemented later) will store the string to int relationship in postgres, but for now this just uses the mock dummy indexer to actually do the conversion. 

In this PR the consumer consumes messages from the `ingest-metrics` topic, translate the payload to have ints instead of strings and then produce to the `snuba-metrics` topic so that snuba can then store the data. 

**RedisMockIndexer**
The temporary redis indexer can be used by changing the following in the `conf/server.py`:

```python
SENTRY_METRICS_INDEXER = "sentry.sentry_metrics.indexer.redis_mock.RedisMockIndexer"
```
It also uses a `bulk_record` method to be able to get and set all the strings (metric name, tag keys and values) for a message at once.

MeredithAnya commented on src/sentry/sentry_metrics/indexer/mock.py:
```diff
@@ -32,7 +32,17 @@ def record(self, organization: Organization, use_case: UseCase, string: str) ->
+            resolved = _STRINGS[string]
+        except KeyError:
+            # just generating a random number for now
+            resolved = random.randint(20, 100)
```
I just added this in for now so that I could test without having to add every single string to the `_STRINGS` dictionary

MeredithAnya commented:
@fpacifici @jjbayer (cc @jan-auer): some questions/concerns/thoughts I have:

* I'm not sure if we want to use the `BatchKafkaConsumer`, or if we should be writing our own Consumer, like what is done for the [`QuerySubscriptionConsumer`](https://github.com/getsentry/sentry/blob/master/src/sentry/snuba/query_subscription_consumer.py#L48). 
* I don't know exactly how best to have the dummy indexer work for what product needs to build on top of this, prior adding in the actual indexer. 
* Is the `UseCase` actually needed for the indexer? Or is that just information that needs to be passed along later to the metrics product data model?

MeredithAnya commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,57 @@
+    def process_message(self, message):
+        original: Dict[str, Any] = json.loads(message.value())
+        new_message = original.copy()
+        project_id = int(new_message["project_id"])
```
oop I think this needs to be the `org_id` instead now

jjbayer commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,57 @@
+        new_message = original.copy()
+        project_id = int(new_message["project_id"])
+
+        metric_id = indexer.resolve(project_id, UseCase.METRIC, new_message["name"])
```
Shouldn't all calls to `indexer.resolve` go to `indexer.record` here? Because we want a new entry to be created when the metric / tag / value appears for the first time.

We would have to change the mock implementation of `record` to not throw when the name is one of the known strings, though.

jjbayer commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,57 @@
+        new_message["retention_days"] = 90
+        return new_message
+
+    def flush_batch(self, batch):
```
I wonder if we can reduce the number of calls to the indexer by doing the transformation per-batch rather than per-message (collecting all strings in the batch and filtering out duplicates).
On the other hand, if the actual string indexer has an in-memory cache, that may not be necessary.

jjbayer commented:
> * I don't know exactly how best to have the dummy indexer work for what product needs to build on top of this, prior adding in the actual indexer.

For release health we cannot really mock tag values, because the `release` tag may have any value. Maybe we can use a redis key-value lookup as the simplest possible indexer implementation?

> * Is the `UseCase` actually needed for the indexer? Or is that just information that needs to be passed along later to the metrics product data model?

I think it's not necessary from a functional perspective -- but maybe for partitioning?

MeredithAnya commented:
> > * I don't know exactly how best to have the dummy indexer work for what product needs to build on top of this, prior adding in the actual indexer.
> 
> For release health we cannot really mock tag values, because the `release` tag may have any value. Maybe we can use a redis key-value lookup as the simplest possible indexer implementation?
> 
> > * Is the `UseCase` actually needed for the indexer? Or is that just information that needs to be passed along later to the metrics product data model?
> 
> I think it's not necessary from a functional perspective -- but maybe for partitioning?

**Updates:**
* Added redis per @jjbayer's suggestion, assuming that it's an okay implementation of it to get things unblocked
* Changed the interface to have only `record` and `reverse_resolve` where `record` does both the work of looking up the value and recording it if it's not already there

fpacifici commented on src/sentry/runner/commands/run.py:
```diff
@@ -504,3 +504,20 @@ def ingest_consumer(consumer_types, all_consumer_types, **options):
+@click.option("--topic", default="ingest-metrics", help="Topic to get subscription updates from.")
+@batching_kafka_options("metrics-consumer")
+@configuration
+def metrics_consumer(**options):
```
Is this going to start the consumer by default? I don't think we need it yet by default.

MeredithAnya replied:
I've had to manually run `sentry run metrics-consumer` so I don't think this starts by default

lynnagara replied:
How come we are adding a separate command rather than using the single `ingest-consumer` command which runs the rest of the ingest consumers? We could still temporarily omit from `--all-consumer-types` and only run it if explicitly called with the metrics consumer type if was the concern. Curious if there is another reason.

MeredithAnya replied:
@lynnagara (cc @fpacifici) the `ingest-consumer` commands ends up using the `IngestConsumerWorker` where as we want to use the `MetricsIndexerWorker` instead. I felt like it was easier to keep these separate for now then to refactor the `ingest-consumer` command. It seems like this could be easily changed down the line if we wanted, but open to changing it now if people feel strongly

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,57 @@
+
+from sentry.sentry_metrics import indexer
+from sentry.sentry_metrics.indexer.base import UseCase
+from sentry.utils import json, kafka_config
```
Please ensure you use rapidjson for parsing and serializing json

https://github.com/getsentry/sentry/blob/master/src/sentry/utils/json.py#L98-L115

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,57 @@
+        new_message = original.copy()
+
+        org_id = int(new_message["org_id"])
+        metric_id = indexer.record(org_id, UseCase.METRIC, new_message["name"])
```
Could we have a `batch_resolve` method? We are always going to do these in batches as each message has many. BOth when writing on the DB or reading from cache, doing it in batches is likely going to be more efficient.

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,57 @@
+class MetricsIndexerWorker(AbstractBatchWorker):
+    def process_message(self, message):
+        original: Dict[str, Any] = json.loads(message.value())
+        new_message = original.copy()
```
Do we actually need to copy the message?

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,57 @@
+        global snuba_metrics
+        for message in batch:
+            cluster_name = snuba_metrics["cluster"]
+            snuba_metrics_publisher = KafkaPublisher(
+                kafka_config.get_kafka_producer_cluster_options(cluster_name),
+                asynchronous=False,
+            )
+            snuba_metrics_publisher.publish(snuba_metrics["topic"], json.dumps(message))
```
This would flush the messages for every message, which is not ideal considering you have batches of messages and you can flush before committing on the ingest topic only once per batch of messages.
We should also set the callback for each message.

fpacifici replied:
Some follow up on what a consumer should do:

When writing a consumer like this that needs to achieve a high throughput, there are a few elements to take into account and a few requirements:
- We cannot auto commit, because we would be committing before processing the message. If there is an error in processing the message is lost. This is taken into account as we do not auto commit
- We don't need to commit after every offset. Less load on the broker and network
- We must not loose messages. So if an error happens at any point during processing we cannot commit the entire batch but only the portion that we sent to Kafka
- producing is asynchronous. So we need to wait for the callback before being sure that the message is persisted in a kafka topic https://docs.confluent.io/clients-confluent-kafka-python/current/overview.html#asynchronous-writes
- We should really avoid duplicates, as they would not be deduplicated in Clickhouse since metrics are stored in a pre-aggregated way. So if there is an error during processing that causes the consumer to crash we should have committed up to the last acknowledged message (acknowledged meaning that we did receive the callback from the producer).
- We should keep producing asynchronously and not flush every message individually as that would have a real impact on throughput.
- Exactly one semantics is technically not achievable as nobody can deduplicate messages. But we can make duplication extremely rare (basically only in case Kafka commit fails multiple times or the consumer crashes for out of memory without being able to commit and after flushing.

So there are a few ways to do that:
- Simple batching consumer. Do the processing phase, then the batch flush sends all the messages, at the end flushes and waits for callbacks. It only commits on kafka the offset of the last callback received.
- Something like the cdc link above. Keep producing messages as soon as they are processed and periodically commit the last offset we got the callback for.

fpacifici reviewed (commented):
Thanks, this is a step in the right direction. 
Will review again tomorrow with some more details on how to produce in an efficient manner. 
I am not sure about the goal of that pubsub class that does not allow us to set a callback.

jjbayer commented on src/sentry/sentry_metrics/indexer/mock.py:
```diff
@@ -1,41 +1,44 @@
+        """
+        client = get_client()
 
-    def resolve(self, organization: Organization, use_case: UseCase, string: str) -> Optional[int]:
```
Contrary to what I said previously, I think it would make sense to keep the `resolve` method. It's already in use [here](https://github.com/getsentry/sentry/blob/e59971700f9f1c1778030cc3944805d845a79e78/src/sentry/snuba/metrics.py#L422), and if the indexer entries ever get a TTL, it would not make sense to prolong the retention every time the indexer is queried from the product side.

jjbayer reviewed (commented):
This looks great! I did a manual end-to-end test locally and everything works as expected.

jjbayer commented on src/sentry/sentry_metrics/indexer/mock.py:
```diff
@@ -1,41 +1,44 @@
 
 from .base import StringIndexer, UseCase
 
-_STRINGS = {
```
Some unit tests actually rely on these mock integer values (e.g. [here](https://github.com/getsentry/sentry/blob/e59971700f9f1c1778030cc3944805d845a79e78/tests/sentry/snuba/test_metrics.py#L106)), so it might make sense to keep the `MockIndexer` as it was, and put the redis-based indexer in a new subclass of `StringIndexer`.

jjbayer commented on src/sentry/sentry_metrics/indexer/redis_mock.py:
```diff
@@ -0,0 +1,150 @@
+
+
+def get_int(string: str) -> int:
+    return int.from_bytes(hashlib.md5(string.encode("utf-8")).digest(), "big") % (10 ** 8)
```
Another nit, but you probably just want to take 8 bytes, right?

```suggestion
    return int.from_bytes(hashlib.md5(string.encode("utf-8")).digest()[:8], "big")
```

nikhars commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,82 @@
+        global snuba_metrics
+        messages = 0
+        cluster_name = snuba_metrics["cluster"]
+        snuba_metrics_producer = Producer(
```
Instead of creating a new Producer on every flush, it would be better if you created it during initialization and kept using it here.

nikhars commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,82 @@
+
+    def shutdown(self) -> None:
+        # do any other processes need to be shutdown?
+        return
```
Do you need to do producer.close() during shutdown?

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,82 @@
+
+
+class MetricsIndexerWorker(AbstractBatchWorker):  # type: ignore
+    def process_message(self, message: Any) -> Any:
+        parsed_message: Dict[str, Any] = json.loads(message.value(), use_rapid_json=True)
```
Can you use a more abstract `Mapping[str, Any]` or `MutableMapping[str, Any]` ?

lynnagara replied:
You have defined parsed_message as `Dict[str, Any]` but returning `Any`. How about a TypedDict here?

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,82 @@
+        metric_name = parsed_message["name"]
+        tags = parsed_message["tags"]
+
+        strings = {metric_name}
+        strings.update(tags.keys())
+        strings.update(tags.values())
```
nit: this can be more compact:
```
strings = {
  metric_name,
  *tags.keys(),
  *tags.values()
}
```

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,82 @@
+        strings.update(tags.keys())
+        strings.update(tags.values())
+
+        mapping = indexer.bulk_record(org_id, list(strings))  # type: ignore
```
Why do we need to ignore the type checking ?

MeredithAnya replied:
when I run `make backend-typing` I get:
`src/sentry/sentry_metrics/indexer/indexer_consumer.py:48: error: Module has no attribute "bulk_record`

I couldn't figure out how to fix it 😞

fpacifici replied:
mh ok. I think it has to do with the way we expose methods with the service delegator.

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,82 @@
+
+        mapping = indexer.bulk_record(org_id, list(strings))  # type: ignore
+
+        new_tags = {}
+        for tag_k, tag_v in tags.items():
+            new_k = mapping[tag_k]
+            new_v = mapping[tag_v]
+            new_tags[new_k] = new_v
```
```
new_tags = {mapping[k]: mapping[v] for k,v in tags.items()}
```

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,82 @@
+
+        parsed_message["tags"] = new_tags
+        parsed_message["metric_id"] = mapping[metric_name]
+        parsed_message["retention_days"] = 90
```
Where is this supposed to come from later? @jan-auer

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,82 @@
+        global snuba_metrics
+        messages = 0
+        cluster_name = snuba_metrics["cluster"]
+        snuba_metrics_producer = Producer(
+            kafka_config.get_kafka_producer_cluster_options(cluster_name),
+        )
```
Please do not rebuild the consumer at each flush, you can create it in the constructor and reuse it, it just not needed. Not sure about the performance impact.

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,82 @@
+            snuba_metrics_producer.produce(
+                topic=snuba_metrics["topic"],
+                key=None,
+                value=json.dumps(message),
```
We generally produce binary strings instead of the python 3 unicode. Please encode the string first.
```
json.dumps(message).encode()
```

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,82 @@
+                value=json.dumps(message),
+                on_delivery=self.callback,
+            )
+            messages += snuba_metrics_producer.poll(0.5)
```
Not sure poll is what you need here. Poll will run the callbacks ready within the timeout 
https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#confluent_kafka.Producer.poll

You may want to ensure all the messages have been sent before committing. Thus you need a call to `flush`.
We can make this better with the streaming consumer when arroyo will be available in Sentry.

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,82 @@
+            raise Exception("didn't get all the callbacks")
+
+    def shutdown(self) -> None:
+        # do any other processes need to be shutdown?
```
No, you can remove this comment.

fpacifici commented on src/sentry/sentry_metrics/indexer/base.py:
```diff
@@ -17,17 +16,20 @@ class StringIndexer(Service):  # type: ignore
+    __all__ = ("record", "resolve", "reverse_resolve", "bulk_record")
 
-    def record(self, organization: Organization, use_case: UseCase, string: str) -> int:
+    def bulk_record(self, org_id: str, strings: List[str]) -> Dict[str, int]:
```
org id is integer.

fpacifici commented on src/sentry/sentry_metrics/indexer/redis_mock.py:
```diff
@@ -0,0 +1,150 @@
+            redis_key_values[string_key] = int_value
+            redis_key_values[int_key] = string
+
+        get_client().mset(redis_key_values)
```
Please set an expiry time. If this is used in the production environment while we develop, we do not want these keys to be lingering around.

lynnagara commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,82 @@
+        for message in batch:
+            snuba_metrics_producer.produce(
+                topic=snuba_metrics["topic"],
+                key=None,
```
Are you sure we don't want to produce messages with any key here?

MeredithAnya replied:
I wasn't sure what we wanted to use as the key, I guess we can just use the `project_id` for now? cc @fpacifici

lynnagara replied:
The key determines the partitioning so we probably don't want a bare `project_id` unless we actually want the topic to be semantically partitioned and are ok with some partitions being a lot larger than others as projects vary in size a lot. I believe the default partitioner will use round robin if there is no key.

fpacifici replied:
round robin is good for this topic

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,82 @@
+    return create_batching_kafka_consumer({topic}, worker=MetricsIndexerWorker(), **options)
+
+
+class MetricsIndexerWorker(AbstractBatchWorker):  # type: ignore
```
Please let's have a test case for the worker. This is going to be the real thing, not a mock.

fpacifici reviewed (changes requested):
Thanks. I think this is close. Only  some specific comments to address, and please add a test for the worker. 

Also I think we should stick to memcached at this point. Redis will be useful when we will be doing fancy things like serving ids from redis or ensuring we hit postgres only once per metrics. Till then memcached is cheaper and scales better (multi threaded).

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,82 @@
+            messages += snuba_metrics_producer.poll(0.5)
+
+        if messages < len(batch):
+            # TODO(meredith): We are not currently keeping track of
+            # which callbacks failed. This means could potentially
+            # be duplicating messages since we don't commit offsets
+            # unless all the callbacks are successful.
+            #
+            # In the future if we know which callback failed, we can
+            # commmit only up to that point and retry on the remaining
+            # messages.
```
Some remarks about this case. The TODO is ok for now. This will be easier to achieve in Arroyo.
Though the scenarios you will need to deal with are:
- Error during flush: We need to ensure to keep track of the offset we reached and positively confirmed for every partition we have assigned. Which also means double checking on how the callback work on the Kafka api. Is it even possible that one message fails and the next succeeds for a given partition? I suspect librdkafka would not send the following one. Though this may happen across partitions.

A few cases are already clean and do not require anything else:
- rebalancing: If your partitions get revoked, the callback on the batching consumer will flush the batch.
- error during processing. In this case the batch was not sent to Kafka so we can reprocess it from scratch.

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,84 @@
+                on_delivery=self.callback,
+            )
+
+        messages_left = self.__producer.flush()
```
Please provide a timeout. We can start with 5 seconds

fpacifici commented on src/sentry/sentry_metrics/indexer/indexer_consumer.py:
```diff
@@ -0,0 +1,84 @@
+            # In the future if we know which callback failed, we can
+            # commmit only up to that point and retry on the remaining
+            # messages.
+            raise Exception("didn't get all the callbacks")
```
Please add some extra data to the exception with the amount of callbacks that were missing.

fpacifici reviewed (commented):
Almost ok, please see my last comment and Nikhar's comment then I will accept

nikhars commented:
LGTM. Nice job.

fpacifici reviewed (approved):
Good first step. Thanks

fpacifici commented:
Please figure out what is wrong with CI before merging
